uber/site_sections/preregistration.py

```diff
@@ -191,11 +191,28 @@ def banned(self, id):
         attendee, group = self._get_unsaved(id)
         return {'attendee': attendee}
 
+    def process_free_prereg(self, session):
+        charge = Charge(listify(self.unpaid_preregs.values()))
+        if charge.total_cost <= 0:
+            for attendee in charge.attendees:
+                attendee.paid = c.NEED_NOT_PAY
+                session.add(attendee)
+
+            for group in charge.groups:
+                session.add(group)
+
+            self.unpaid_preregs.clear()
+            self.paid_preregs.extend(charge.targets)
+            raise HTTPRedirect('paid_preregistrations?payment_received={}', charge.dollar_amount)
+        else:
+            message = "These badges aren't free! Please pay for them."
+            raise HTTPRedirect('index?message={}', message)
+
     @credit_card
     def prereg_payment(self, session, payment_id, stripeToken):
         charge = Charge.get(payment_id)
         if not charge.total_cost:
-            message = 'Your preregistration has already been paid for, so your credit card has not been charged'
+            message = 'Your total cost was $0. Your credit card has not been charged.'
```

Comment:
By the way, about redirecting to process_free_prereg -- while that would've been slightly better UX, I think this is a rare enough occurrence that it's not worth sweating. You could even argue that it's better to explicitly tell people their card wasn't charged and then make them free-register instead of doing it all invisibly.